src/ai/llm/messages/manager.ts

```diff
@@ -226,42 +272,82 @@ export class MessageManager {
 
-        let total = 0;
-        const overheadPerMessage = 4; // Approximation for message format overhead
+        let totalTokens = 0;
 
         try {
-            // Count system prompt
-            if (this.systemPrompt) {
-                total += this.tokenizer.countTokens(this.systemPrompt);
-            }
-
-            // Count each message
-            for (const message of this.history) {
-                if (message.content) {
-                    total += this.tokenizer.countTokens(message.content);
+            // Get the full message history formatted for the target LLM provider
+            const formattedMessages = this.formatter.format(this.history, this.systemPrompt);
+
+            // Iterate through the provider-specific formatted messages
+            for (const formattedMessage of formattedMessages) {
+                // --- Token Counting Logic based on potential Vercel/CoreMessage structure --- 
+                // This assumes the formatter returns something close to CoreMessage[] 
+                // Adjust if your specific formatter returns a different structure
+                
+                if (formattedMessage?.content) {
+                    if (typeof formattedMessage.content === 'string') {
+                        totalTokens += this.tokenizer.countTokens(formattedMessage.content);
+                    } else if (Array.isArray(formattedMessage.content)) {
+                        // For multimodal, currently count only text parts
+                        formattedMessage.content.forEach((part: any) => {
+                            if (part.type === 'text' && typeof part.text === 'string') {
+                                totalTokens += this.tokenizer.countTokens(part.text);
+                            }
+                            // NOTE: Image token counting is complex and omitted for now.
+                        });
+                    }
                 }
 
-                // Count tool calls if present
-                if (message.toolCalls) {
-                    for (const call of message.toolCalls) {
-                        total += this.tokenizer.countTokens(call.function.name);
-                        // Ensure arguments exist and are strings before counting
-                        if (call.function.arguments) {
-                            total += this.tokenizer.countTokens(call.function.arguments);
-                        }
-                    }
+                // Count tokens for tool calls if present in the formatted message
+                // Vercel SDK uses 'toolCalls' or 'tool_calls' depending on context/version
+                const toolCalls = formattedMessage?.toolCalls || formattedMessage?.tool_calls;
+                if (toolCalls) {
+                     // Stringify to approximate token cost, as structure can vary
+                    const toolCallsString = JSON.stringify(toolCalls);
+                    totalTokens += this.tokenizer.countTokens(toolCallsString);
```

Comment:
do we need this change in logic for counting tokens in tool calls? 

also we should stick to `toolCalls` or `tool_calls` and not both - whichever one is the one our version of vercel sd actually uses. if vercel's name changes in a new version we can bump up our version as well and change this - no need to support both